File: cpp/Ice/forwarder/Forwarder.cpp
// Copyright (c) ZeroC, Inc.

#include "Forwarder.h"

using namespace std;

ForwardingServer::Forwarder::Forwarder(Ice::ObjectPrx targetTemplate)
    : _targetTemplate{std::move(targetTemplate)}
{
}

void
ForwardingServer::Forwarder::dispatch(Ice::IncomingRequest& request, function<void(Ice::OutgoingResponse)> sendResponse)
{
    // Make a copy of the current object carried by the request.
    const Ice::Current current{request.current()};

    // Create a proxy with the the desired identity and facet.
    Ice::ObjectPrx target = _targetTemplate.ice_identity(current.id).ice_facet(current.facet);

    if (current.requestId == 0)
    {
        // The incoming request is one-way, so we reconfigure target to be one-way.
        target = target.ice_oneway();
    }

    // Read the encapsulation containing the in-parameters.
    const std::byte* inEncapsulationStart = nullptr;
    int32_t inEncapsulationSize = 0;
    request.inputStream().readEncapsulation(inEncapsulationStart, inEncapsulationSize);

    // Make the invocation asynchronously. This call reports most exceptions through its exception callback.
    target.ice_invokeAsync(
        current.operation,
        current.mode,
        make_pair(inEncapsulationStart, inEncapsulationStart + inEncapsulationSize),
        [sendResponse, current](bool ok, pair<const std::byte*, const std::byte*> outEncapsulation)
        {
            // The response callback, executed by a thread from the Ice client thread pool when the invocation
            // completes successfully (ok is true) or with a user exception (ok is false).
            // We create an OutgoingResponse object and send it back to the client with sendResponse.
            sendResponse(Ice::makeOutgoingResponse(ok, outEncapsulation, current));
        },
        [sendResponse, current](std::exception_ptr exceptionPtr)
        {
            // The exception callback.
            // We create an OutgoingResponse object with the exception and send it back to the client with
            // sendResponse. If the exception is an Ice local exception that cannot be marshaled, such as
            // Ice::ConnectionRefusedException, makeOutgoingResponse marshals an Ice::UnknownLocalException.
            sendResponse(Ice::makeOutgoingResponse(exceptionPtr, current));
        },
        nullptr, // no sent callback
        current.ctx);
}

File: java/Ice/forwarder/forwardingserver/src/main/java/com/example/ice/forwarder/forwardingserver/Forwarder.java
// Copyright (c) ZeroC, Inc.

package com.example.ice.forwarder.forwardingserver;

import com.zeroc.Ice.IncomingRequest;
import com.zeroc.Ice.ObjectPrx;
import com.zeroc.Ice.OutgoingResponse;
import com.zeroc.Ice.UserException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/**
 * Forwarder is an Ice servant that implements com.zeroc.Ice.Object by forwarding all requests it receives to a remote
 * Ice object.
 */
class Forwarder implements com.zeroc.Ice.Object {
    // A template for the target proxy.
    private final ObjectPrx _targetTemplate;

    // Implements the abstract method dispatch defined on com.zeroc.Ice.Object.
    @Override
    public CompletionStage<OutgoingResponse> dispatch(IncomingRequest request) throws UserException {
        // Create a proxy with the desired identity and facet.
        ObjectPrx target = _targetTemplate.ice_identity(request.current.id).ice_facet(request.current.facet);

        if (request.current.requestId == 0) {
            // The incoming request is one-way, so we reconfigure target to be one-way.
            target = target.ice_oneway();
        }

        // Make the invocation asynchronously.
        CompletableFuture<com.zeroc.Ice.Object.Ice_invokeResult> future = target.ice_invokeAsync(
            request.current.operation,
            request.current.mode,
            request.inputStream.readEncapsulation(null), // the in-parameters encapsulation
            request.current.ctx);

        // Upon success, create an OutgoingResponse from the invokeResult. If the invocation fails with an exception,
        // the object adapter calling this method will convert the exception into an UnknownLocalException or
        // UnknownException.
        return future.thenApply(
            invokeResult -> request.current.createOutgoingResponse(invokeResult.returnValue, invokeResult.outParams));
    }

    // Constructs a forwarder.
    Forwarder(ObjectPrx targetTemplate) {
        _targetTemplate = targetTemplate.ice_twoway();
    }
}
